File: pfutils-cpp/src/FileManager.cpp
#include "FileManager.h"
#include <utility>
#include <fstream>

FileManager::FileManager(
    uint32_t num_threads, 
    uint32_t chunk_size
) : bulk_pool(num_threads, chunk_size), copy_options(std::filesystem::copy_options::update_existing)
{}

void FileManager::copy_file(
    std::filesystem::path &src,
    std::filesystem::path &dst
) {
    this->bulk_pool.push(
        [src, dst, copy_options=copy_options]()->void{
            std::filesystem::copy_file(src, dst, copy_options);
        }
    );
}

void FileManager::remove_file(
    std::filesystem::path &path
) {
    this->bulk_pool.push(
        [path]()->void{
            std::filesystem::remove(path);
        }
    );
}

void FileManager::write_file(
    std::filesystem::path &path,
    uint32_t file_size,
    bool randomness
) {
    this->bulk_pool.push(
        [path, file_size]()->void{
            char *byte_array = "0123456789abcdef";
            std::ofstream file(path, std::ios::binary | std::ios::out);

            for(uint32_t i=0; i<64 * file_size; ++i) {
                file.write(byte_array, 16);
            }
        }
    );
}

std::shared_ptr<FutureVector<void>> FileManager::flush_and_iter() {
    return this->bulk_pool.flush_and_iter();
}

File: pfutils/utils/file_manager/file_manager.py
import os
import shutil
from pfutils.utils.bulk_pool import ProcessPoolExecutor

def copy_func(src, dst):
    try:
        return shutil.copy2(src, dst)
    except PermissionError as e:
        raise FileSystemError(f'cannot write file [{e.filename}]')
    except FileNotFoundError as e:
        raise FileSystemError(f'cannot copy file [{e.filename}]')

def write_func(path, file_size, randomness):
    with open(path, 'wb') as f:
        f.write(b'0123456789abcdef' * 64 * file_size)
    
class FileManager:
    def __init__(self, num_workers, chunk_size):
        self.pool = ProcessPoolExecutor(num_workers, chunksize=chunk_size)
        self.futures = []

    def copy_file(self, src, dst):
        self.pool.submit(copy_func, src, dst)
    
    def remove_file(self, path):
        self.pool.submit(os.remove, path)

    def write_file(self, path, file_size, randomness):
        self.pool.submit(write_func, path, file_size, randomness)

    def flush_and_iter(self):
        return self.pool.flush()

    def __enter__(self):
        return self

    def __exit__(self, *args, **kwargs):
        pass

class FileSystemError(Exception):
    def __init__(self, msg):
        super(Exception, self).__init__(msg)
    
    def __str__(self):
        return f'filesystem error: {super().__str__()}'